### PR TITLE
chore: move to CircleCI and fix codecov

### PR DESCRIPTION
**Pull request by JustinBeckwith**
This uses js-green-licenses as an example, and moves us over to CircleCI. Feedback welcomed!

cc @alexander-fenster

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+version: 2.0
+workflows:
+  version: 2
+  tests:
+    jobs:
+      - node4:
+          filters:
+            tags:
+              only: /^v\d+(\.\d+){2}(-.*)?$/
```
Comment (jinwoo) [inline .circleci/config.yml]: I learned a trick from @kjin for defining this regexp only once. https://github.com/google/node-h2-auto-push/blob/9b09ae27c31a637ae32a23c05957944a5d530ef6/.circleci/config.yml#L2

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+  publish_npm:
+    docker:
+      - image: node:8
+        user: node
+    steps:
+      - checkout
+      - run:
+          name: Set npm authentication.
+          command: echo "//registry.npmjs.org/:_authToken=${NPM_TOKEN}" > ~/.npmrc
```
Comment (jinwoo) [inline .circleci/config.yml]: If you want npm auto-publish on version tags, you need to add the NPM_TOKEN env var to the CircleCI settings. The token is in valentine.

File: test/test.googleauth.ts [hunk 3]
```diff
@@ -1176,7 +1176,8 @@ describe('.getApplicationDefault', () => {
        // tslint:disable-next-line no-any
        assert.equal(null, (result3 as any).specialTestBit);
        assert.notEqual(cachedCredential, result3);
-     });
+     })
+      .timeout(5000);
```
Comment (jinwoo) [inline test/test.googleauth.ts]: Curious. Why is this needed?

File: .circleci/config.yml (hunk 1 above)
Comment (alexander-fenster) [reply .circleci/config.yml]: Why do we need to filter by tag for these tasks at all?

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      - run:
+          name: Install modules and dependencies.
+          command: npm install
+      - run:
+          name: Publish the module to npm.
+          command: npm publish
```
Comment (alexander-fenster) [inline .circleci/config.yml]: To make CI run for all commits, but publish only for v.* tags, all tasks like node4, node6, etc. should have a filter /.*/, but the publish task should be restricted to v.*.  See example here: https://github.com/googleapis/nodejs-os-login/blob/master/.circleci/config.yml - we run tests on all versions for all commits, run lint and docs for all commits, but run publish only on v* tags.

File: .circleci/config.yml (hunk 1 above)
Comment (jinwoo) [reply .circleci/config.yml]: This is so that these tests also run for those tags as well as changes in branches. That's needed because `node4`, `node6`, etc. are requirements of `publish_npm`, which is triggered only for such tags.

CircleCI build is performed for branches by default, but is not for tags by default.

File: .circleci/config.yml (hunk 4 above)
Comment (jinwoo) [reply .circleci/config.yml]: No. The build runs for branches by default, so you don't need a filter `/.*/`. `node-h2-auto-push` has the same configuration and npm publish works fine: https://github.com/google/node-h2-auto-push/blob/master/.circleci/config.yml

**Comment by alexander-fenster** [comment]
Also, in circleci -> project settings -> Advanced settings, enable builds for forked repos, that way, you'll get CI run for this PR as well.

File: .circleci/config.yml (hunk 4 above)
Comment (jinwoo) [reply .circleci/config.yml]: See this doc: https://github.com/google/node-h2-auto-push/blob/master/.circleci/config.yml

Comment (jinwoo) [reply .circleci/config.yml]: Oops. Sorry. See this doc: https://circleci.com/docs/2.0/workflows/#git-tag-job-execution

File: .circleci/config.yml (hunk 1 above)
Comment (alexander-fenster) [reply .circleci/config.yml]: hmm. we have different setup in our other repos (see my comment below)

File: .circleci/config.yml (hunk 4 above)
Comment (alexander-fenster) [reply .circleci/config.yml]: hmm. interesting
probably it's because we also have extra tasks (lint, docs, and system and sample tests for some repos) that depend on the branches?

Comment (alexander-fenster) [reply .circleci/config.yml]: so what happens in our case is, {node4,node6,node7,node8,node9} -> {lint,docs} runs for all commits, but {node4,node6,node7,node8,node9} -> {lint,docs} -> publish_npm runs for tagged only, hence we probably need to have a filter of /.*/ on all tasks and a filter of /v.*/ on publish_npm.

File: test/test.googleauth.ts (hunk 3 above)
Comment (JustinBeckwith) [reply test/test.googleauth.ts]: This test will randomly time out from time to time.  I am dropping it from this PR and doing another.

File: .circleci/config.yml (hunk 1 above)
Comment (JustinBeckwith) [reply .circleci/config.yml]: Woah this is rad.  Thanks!

**Comment by JustinBeckwith** [comment]
@jinwoo I mentioned green licenses because I copied your config.yml :)

File: .circleci/config.yml (hunk 2 above)
Comment (JustinBeckwith) [reply .circleci/config.yml]: 👍 will do this as soon as this lands

Comment (JustinBeckwith) [reply .circleci/config.yml]: Which secret in valentine am I looking for?  Can you make sure I have access?

Comment (jinwoo) [reply .circleci/config.yml]: It's stored as "google-node-team@npm's NPM_TOKEN". I think you have access. Let me know if you don't.
